### ISSUE TITLE
Connect() throwing stack overflow error 

### ISSUE THREAD

**Issue by ramonmpa** [robinrodricks/FluentFTP#808]
**FTP OS:** Unix

**FTP Server:** Vsftpd

**Computer OS:** docker application (using base image mcr.microsoft.com/dotnet/runtime:5.0)

**FluentFTP Version:** 35.2.2

Applications runs normally but after some time when method FluentFTP .Connect() is called its crashing entire application with "stack overflow" error. The error log has more than 14.500 calls to these methods:
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()

Looking at source code of FtpClient_Connection.cs and FtpClient_Stream.cs I could see that this recursive call trace exists in the logic: .Connect() -> .Authenticate() -> .Execute() -> .Connect() ....

I don't know what exactly is happening in the server for FluentFTP starts this endless recursive calls.

Any idea what could cause this behaviour in FluentFTP?

Thanks.

Code implementation to connect:
```
public IFtpClient Connect(IFtpClient ftpClient)
{
	var credentials = new NetworkCredential("test", "test");
	ftpClient = new FtpClient("test", 21, credentials);
	ftpClient.EncryptionMode = FtpEncryptionMode.Explicit;
	ftpClient.DataConnectionType = FtpDataConnectionType.AutoPassive;
	ftpClient.DataConnectionEncryption = true;
	ftpClient.SslProtocols = System.Security.Authentication.SslProtocols.None;
	ftpClient.ValidateCertificate += new FtpSslValidation(OnValidateCertificate);
	var cer = new X509Certificate2(Path.GetFullPath("./vsftpd.crt"));
	ftpClient.ClientCertificates.Add(cer);
	ServicePointManager.ServerCertificateValidationCallback = ServerCertificateValidationCallback;
	ftpClient.Connect();
	return ftpClient;
}

private void OnValidateCertificate(FtpClient control, FtpSslValidationEventArgs e)
{
	// add logic to test if certificate is valid here
	e.Accept = true;
}
private bool ServerCertificateValidationCallback(object sender,
										X509Certificate certificate,
										X509Chain chain,
										SslPolicyErrors sslPolicyErrors)
{
	return true;
}
```

**Logs :**

```
....

# Connect()
Status:   Connecting to ***:21
Response: 220 (vsFTPd 3.0.3)
Status:   Detected FTP server: VsFTPd
Command:  AUTH TLS
Response: 234 Proceed with negotiation.
Status:   FTPS Authentication Successful
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0.0055322.
Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
Status:   Disposing FtpSocketStream...

# Connect()
Status:   Connecting to ***:21
Response: 220 (vsFTPd 3.0.3)
Status:   Detected FTP server: VsFTPd
Command:  AUTH TLS
Response: 234 Proceed with negotiation.
Status:   FTPS Authentication Successful
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0.0047787.
Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
Status:   Disposing FtpSocketStream...

# Connect()
Status:   Connecting to ***:21
Response: 220 (vsFTPd 3.0.3)
Status:   Detected FTP server: VsFTPd
Command:  AUTH TLS
Response: 234 Proceed with negotiation.
Status:   FTPS Authentication Successful
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0.005132.
Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
Status:   Disposing FtpSocketStream...

# Connect()
Status:   Connecting to ***:21
Response: 220 (vsFTPd 3.0.3)
Status:   Detected FTP server: VsFTPd
Command:  AUTH TLS
Response: 234 Proceed with negotiation.
Status:   FTPS Authentication Successful
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0.0049426.
Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
Status:   Disposing FtpSocketStream...

# Connect()
Status:   Connecting to ***:21
Response: 220 (vsFTPd 3.0.3)
Status:   Detected FTP server: VsFTPd
Command:  AUTH TLS
Response: 234 Proceed with negotiation.
Status:   FTPS Authentication Successful
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0.0051661.
Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
Status:   Disposing FtpSocketStream...

# Connect()
Status:   Connecting to ***:21
Response: 220 (vsFTPd 3.0.3)
Status:   Detected FTP server: VsFTPd
Command:  AUTH TLS
Response: 234 Proceed with negotiation.
Status:   FTPS Authentication Successful
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0.0044794.
Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
Status:   Disposing FtpSocketStream...

# Connect()
Status:   Connecting to ***:21
Response: 220 (vsFTPd 3.0.3)
Status:   Detected FTP server: VsFTPd
Command:  AUTH TLS
Response: 234 Proceed with negotiation.
Status:   FTPS Authentication Successful
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0.004704.
Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
Status:   Disposing FtpSocketStream...

.... (+ thousands of #Connect())

Stack overflow.
   at Interop+Ssl.SslDoHandshake(Microsoft.Win32.SafeHandles.SafeSslHandle)
   at Interop+OpenSsl.DoSslHandshake(Microsoft.Win32.SafeHandles.SafeSslHandle, System.ReadOnlySpan`1<Byte>, Byte[] ByRef, Int32 ByRef)
   at System.Net.Security.SslStreamPal.HandshakeInternal(System.Net.Security.SafeFreeCredentials, System.Net.Security.SafeDeleteSslContext ByRef, System.ReadOnlySpan`1<Byte>, Byte[] ByRef, System.Net.Security.SslAuthenticationOptions)
   at System.Net.Security.SecureChannel.GenerateToken(System.ReadOnlySpan`1<Byte>, Byte[] ByRef)
   at System.Net.Security.SecureChannel.NextMessage(System.ReadOnlySpan`1<Byte>)
   at System.Net.Security.SslStream+<ForceAuthenticationAsync>d__171`1[[System.Net.Security.AsyncReadWriteAdapter, System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Security.SslStream+<ForceAuthenticationAsync>d__171`1[[System.Net.Security.AsyncReadWriteAdapter, System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]], System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ForceAuthenticationAsync>d__171`1<System.Net.Security.AsyncReadWriteAdapter> ByRef)
   at System.Net.Security.SslStream.ForceAuthenticationAsync[[System.Net.Security.AsyncReadWriteAdapter, System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](System.Net.Security.AsyncReadWriteAdapter, Boolean, Byte[], Boolean)
   at System.Net.Security.SslStream.ProcessAuthentication(Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Security.SslStream.AuthenticateAsClientAsync(System.Net.Security.SslClientAuthenticationOptions, System.Threading.CancellationToken)
   at System.Net.Security.SslStream.AuthenticateAsClientAsync(System.String, System.Security.Cryptography.X509Certificates.X509CertificateCollection, System.Security.Authentication.SslProtocols, Boolean)
   at FluentFTP.FtpSocketStream.ActivateEncryption(System.String, System.Security.Cryptography.X509Certificates.X509CertificateCollection, System.Security.Authentication.SslProtocols)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   
   -------- +14.000 calls to ----------------
   
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   
   -------- +14.000 calls to ----------------
   
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()
   at FluentFTP.FtpClient.Execute(System.String)
   at FluentFTP.FtpClient.Authenticate(System.String, System.String, System.String)
   at FluentFTP.FtpClient.Connect()


```

**Comment by FanDjango** [comment]
IMO here is what is happening: Your `Connect()` profile causes some additional output in addition to the successful auth negotiation. Causes unexpected stale data on the socket, which then causes the whole thing to begin again.

1. You could try using `AutoConnect()`.

Also I see in `Authenticate(...)`:

```
				// fix for #620: some servers send multiple responses that must be read and decoded,
				// otherwise the connection is aborted and remade and it goes into an infinite loop
```
#620 is not the same topic, but this could still be relevant - this code part handles unexpected (i.e. stale) data.

2. You could try again with a current FluentFTP version. It has come a long way since then.

**Comment by FanDjango** [comment]
Closed on account of no feedback
